add guide for advanced routing (#1871)
solo-io/gloo-mesh: docs/content/guides/gateway/advanced_routing.md

```shell
kubectl get service -n istio-system
```

You should be able to the load balancer's external IP address in the output, in the example below it's `32.12.34.555`:

{{< highlight shell "hl_lines=2" >}}
NAME                   TYPE           CLUSTER-IP      EXTERNAL-IP    PORT(S)                                                                      AGE
istio-ingressgateway   LoadBalancer   10.96.229.177   32.12.34.555   15021:31911/TCP,80:30166/TCP,443:32302/TCP,15012:30471/TCP,15443:31931/TCP   10m
istiod                 ClusterIP      10.96.180.254   <none>         15010/TCP,15012/TCP,443/TCP,15014/TCP                                        10m
{{< /highlight >}}

Alternatively you can forward the port to access locally, which is sufficient for the purposes of this guide:

```shell
kubectl --context $CONTEXT_1 -n istio-system port-forward deploy/istio-ingressgateway 8081
```

### Adding Routes

We are now going to add some routes in order to split the reviews service into separate v1 and v2 service then route to
both. We will start by deleting the existing review service:

```shell
kubectl --context $CONTEXT_1 -n bookinfo delete svc reviews
```

> And adding two services for the v1 and v2 deployments:

Reviewer: We might be able to skip this step of deleting/creating versioned services and instead have users use subsets in their kubeService selector directly in the example? What do you think?

        clusters:
        - cluster-1
        labels:
          istio: ingressgateway
        namespaces:
        - istio-system
  connectionHandlers:
  - http:
      routeConfig:
      - virtualHost:
          domains:
          - www.example.com
          routes:
          - matchers:
            - uri:
                prefix: /ratings
            name: ratings
            routeAction:
              destinations:
              - kubeService:
                  clusterName: cluster-1
                  name: ratings
                  namespace: bookinfo
          - matchers:
            - uri:
                prefix: /reviews/v1
            name: reviews-v1
            routeAction:
              destinations:
              - kubeService:
                  clusterName: cluster-1
                  name: reviews-v1
                  namespace: bookinfo
          - matchers:
            - uri:
                prefix: /reviews/v2
            name: reviews-v2
            routeAction:
              destinations:
              - kubeService:
                  clusterName: cluster-1
                  name: reviews-v2
                  namespace: bookinfo
EOF
{{< /tab >}}
{{< /tabs >}}

> 

Author: Will be adding in some checks once I test everything and finalize the guide

spec:
  routes:
  - matchers:
    - uri:
        prefix: /v1
    name: reviews-v1
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews-v1
          namespace: bookinfo
  - matchers:
    - uri:
        prefix: /v2
    name: reviews-v2
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews-v2
          namespace: bookinfo
{{< /tab >}}
{{< tab name="CLI Inline" codelang="shell" >}}
cat << EOF | kubectl apply --context $CONTEXT_1 -f -
apiVersion: networking.enterprise.mesh.gloo.solo.io/v1beta1
kind: RouteTable
metadata:
  name: reviews-rt
  namespace: gloo-mesh
  labels:
    service: reviews
spec:
  routes:
  - matchers:
    - uri:
        prefix: /v1
    name: reviews-v1
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews-v1
          namespace: bookinfo
  - matchers:
    - uri:
        prefix: /v2
    name: reviews-v2
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews-v2
          namespace: bookinfo
EOF
{{< /tab >}}
{{< /tabs >}}

> Now, we will use delegate actions on the `VirtualGatewy` to send all `/reviews` requests to our newly minted route table:

Reviewer: ```suggestion
Now, we will use delegate actions on the `VirtualGateway` to send all `/reviews` requests to our newly minted route table:
```

{{< tabs >}}
{{< tab name="YAML File" codelang="yaml">}}
apiVersion: networking.enterprise.mesh.gloo.solo.io/v1beta1
kind: RouteTable
metadata:
  name: reviews-misc-rt
  namespace: gloo-mesh
  labels:
    service: reviews
spec:
  routes:
  - matchers:
    - uri:
        exact: /versions
    name: versions
    directResponseAction:
      status: 200
      body: "['v1', 'v2']"
  - matchers:
    - uri:
        prefix: /
    name: not-found
    directResponseAction:
      status: 404
      body: "'not found'"
{{< /tab >}}
{{< tab name="CLI Inline" codelang="shell" >}}
cat << EOF | kubectl apply --context $CONTEXT_1 -f -
apiVersion: networking.enterprise.mesh.gloo.solo.io/v1beta1
kind: RouteTable
metadata:
  name: reviews-misc-rt
  namespace: gloo-mesh
  labels:
    service: reviews
spec:
  routes:
  - matchers:
    - uri:
        exact: /versions
    name: versions
    directResponseAction:
      status: 200
      body: "['v1', 'v2']"
  - matchers:
    - uri:
        prefix: /
    name: not-found
    directResponseAction:
      status: 404
      body: "'not found'"
EOF
{{< /tab >}}
{{< /tabs >}}

> We can add the table to our `VirtualGateway`:

Reviewer 2: ```suggestion
Next, add the table to the `VirtualGateway`.
```

        - cluster-1
        labels:
          istio: ingressgateway
        namespaces:
        - istio-system
  connectionHandlers:
  - http:
      routeConfig:
      - virtualHost:
          domains:
          - www.example.com
          routes:
          - matchers:
            - uri:
                prefix: /ratings
            name: ratings
            routeAction:
              destinations:
              - kubeService:
                  clusterName: cluster-1
                  name: ratings
                  namespace: bookinfo
          - matchers:
            - uri:
                prefix: /reviews
            name: reviews
            routeAction:
              destinations:
              - kubeService:
                  clusterName: cluster-1
                  name: reviews-v1
                  namespace: bookinfo
          - matchers:
            - uri:
                prefix: /reviews
            name: reviews
            delegateAction:
              refs:
              - name: reviews-rt
                namespace: gloo-mesh
              - name: reviews-misc-rt
                namespace: gloo-mesh
EOF
{{< /tab >}}
{{< /tabs >}}

### Improving the Sorting Logic

> In the last section you added two `RouteTables` to handle rouging logic for the reviews service then used a delegate
> action in order to route to them. One thing you may have noticed is that we had to manually add every `RouteTable` to
> the `VirtualGateway` delegate action. We can swap out the list of references for a selector instead:

Reviewer 2: ```suggestion
In the previous section, you added two `RouteTables` to handle routing logic for the reviews service and used a delegate
action to route to the the route tables. This process included adding every `RouteTable` to the `VirtualGateway` delegate 
action. However, depending on your requirements, manually adding individual route table references might not 
be viable. Instead, you can specify a selector to the delegate action.
```

    - kubeServiceMatcher:
        clusters:
        - cluster-1
        labels:
          istio: ingressgateway
        namespaces:
        - istio-system
  connectionHandlers:
  - http:
      routeConfig:
      - virtualHost:
          domains:
          - www.example.com
          routes:
          - matchers:
            - uri:
                prefix: /ratings
            name: ratings
            routeAction:
              destinations:
              - kubeService:
                  clusterName: cluster-1
                  name: ratings
                  namespace: bookinfo
          - matchers:
            - uri:
                prefix: /reviews
            name: reviews
            routeAction:
              destinations:
              - kubeService:
                  clusterName: cluster-1
                  name: reviews-v1
                  namespace: bookinfo
          - matchers:
            - uri:
                prefix: /reviews
            name: reviews
            delegateAction:
              selector:
                namespaces:
                - gloo-mesh
                labels
                  service: reviews
EOF
{{< /tab >}}
{{< /tabs >}}

> Now all `RouteTables` in the gloo-mesh namespace with the `service: reviews` will be automatically delegated to without
> any changes to the `VirtualGateway`. Both manually listing `RouteTable`r and selecting them can be desirable based on
> the requirements of your system. You can even use a combination, but note that `RouteTable`s listed by reference take
> higher precedence than selected ones, which are sorted alphabetically by namespace then name since Kubernetes does not
> guarantee a deterministic order when selecting multiple objects.

Reviewer 2: ```suggestion
Requests are now automatically delegated to all `RouteTables` in the `gloo-mesh` namespace with the `service: reviews` label without
any changes to the `VirtualGateway`.
You can use a combination of specific route table references and label selectors. However, route table references take precedence over 
route table selectors, which are sorted alphabetically by namespace and then name because Kubernetes does not
guarantee a deterministic order when selecting multiple objects.
```

> One thing you may have noticed is our second route table has a catchall 404 route. We need to ensure that this route is
> always last, otherwise it will short circuit our other routes. We can do this by adding weights to our `RouteTable`s.
> Let's give our main reviews `RouteTable` a weight of 10 so all its routes come before the other `RouteTable`, which has
> a default weight of 0.

Reviewer 2: ```suggestion
In the `reviews-misc-rt` route table resource, `not-found` is defined as a catch-all 404 route. This route must always be 
last in the list of routes that requests are matched against to ensure that it does not bypass the other routes for services. 
To designate the order of routes that requests are matched against, you can add weights to each `RouteTable`.
For example, add `weight: 10` to the `reviews-rt` route table so that all its routes are listed before the routes in the `reviews-misc-rt` 
route table, which has a default weight of 0.
```

  routes:
  - matchers:
    - uri:
        prefix: /v1
    name: reviews-v1
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews-v1
          namespace: bookinfo
  - matchers:
    - uri:
        prefix: /v2
    name: reviews-v2
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews-v2
          namespace: bookinfo
{{< /tab >}}
{{< tab name="CLI Inline" codelang="shell" >}}
cat << EOF | kubectl apply --context $CONTEXT_1 -f -
apiVersion: networking.enterprise.mesh.gloo.solo.io/v1beta1
kind: RouteTable
metadata:
  name: reviews-rt
  namespace: gloo-mesh
  labels:
    service: reviews
spec:
  weight: 10
  routes:
  - matchers:
    - uri:
        prefix: /v1
    name: reviews-v1
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews-v1
          namespace: bookinfo
  - matchers:
    - uri:
        prefix: /v2
    name: reviews-v2
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews-v2
          namespace: bookinfo
EOF
{{< /tab >}}
{{< /tabs >}}

> Now its routes will always come first. Alternatively, in situations where the routes are as neatly segmented into
> separate route tables, you can change the delegate action's sort method to `ROUTE_SPECIFICITY`, which will use a
> heuristic to estimate how specific a routes is and then order them such that more specific routes come before more
> general routes. We can enable it on the delegate action:

Reviewer 2: ```suggestion
Routes for services in `reviews-rt` are now first in the route list. Alternatively, when routes are segmented into
separate route tables, you can change the delegate action's sort method to `ROUTE_SPECIFICITY`. This sort method uses a
heuristic to estimate the specificity of each route, and orders routes from most to least specific. To enable 
this sort method, add `sortMethod: ROUTE_SPECIFICITY` to the delegate action.
```

kind: RouteTable
metadata:
  name: demo-routetable
  namespace: gloo-mesh
spec:
  routes:
  - matchers:
    - uri:
        prefix: /v1
    name: v1
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews
          namespace: bookinfo
          subset:
            version: v1
      pathRewrite: /reviews
  - matchers:
    - uri:
        prefix: /v2
    name: v2
    routeAction:
      destinations:
      - kubeService:
          clusterName: cluster-1
          name: reviews
          namespace: bookinfo
          subset:
            version: v2
      pathRewrite: /reviews
EOF
{{< /tab >}}
{{< /tabs >}}

Now, specify a `delegateAction` field in the `VirtualGateway` to send all `/reviews` requests to the reviews route table.

{{< tabs >}}
{{< tab name="YAML File" codelang="yaml" >}}
apiVersion: networking.enterprise.mesh.gloo.solo.io/v1beta1
kind: VirtualHost
metadata:
  name: demo-virtualhost
  namespace: gloo-mesh
spec:
  domains:
  - www.example.com
  routes:
  - matchers:
    - uri:
        prefix: /ratings
    name: ratings
    routeAction:
      destinations:
      - kubeService:

>           clusterName: mgmt-cluster

Reviewer: Might need to do a find/replace of mgmt-cluster -> cluster-1 and remote-cluster -> cluster-2

  - matchers:
    - uri:
        prefix: /ratings
    name: ratings
    routeAction:
      destinations:
      - kubeService:
          clusterName: mgmt-cluster
          name: ratings
          namespace: bookinfo
  - matchers:
    - uri:
        prefix: /reviews
    name: reviews
    delegateAction:
      refs:
      - name: demo-routetable
        namespace: gloo-mesh
      - name: demo-routetable2
        namespace: gloo-mesh
{{< /tab >}}
{{< tab name="CLI Inline" codelang="shell" >}}
cat << EOF | kubectl apply --context $CONTEXT_1 -f -
apiVersion: networking.enterprise.mesh.gloo.solo.io/v1beta1
kind: VirtualHost
metadata:
  name: demo-virtualhost
  namespace: gloo-mesh
spec:
  domains:
  - www.example.com
  routes:
  - matchers:
    - uri:
        prefix: /ratings
    name: ratings
    routeAction:
      destinations:
      - kubeService:
          clusterName: mgmt-cluster
          name: ratings
          namespace: bookinfo
  - matchers:
    - uri:
        prefix: /reviews
    name: reviews
    delegateAction:
      refs:
      - name: demo-routetable
        namespace: gloo-mesh
      - name: demo-routetable2
        namespace: gloo-mesh
EOF
{{< /tab >}}
{{< /tabs >}}

> It is important that the second route table comes second in the list of references, otherwise it will short circuit all

Reviewer: Should we explain the term "short circuit" here? Basically that the routes are parsed in order and if a very generic path like "prefix '/'" is the first route then all subsequent routes can become unreachable?

the other routes. You should now be able to request the reviews services endpoints to get the reviews and all other

> random routes should return our custom `404` message.

Reviewer: How about "all other random routes" -> "all other paths"?

> It is important that the second route table comes second in the list of references, otherwise it will short circuit all
> the other routes. You should now be able to request the reviews services endpoints to get the reviews and all other
> random routes should return our custom `404` message.

Reviewer 2: ```suggestion
It is important that the second route table is listed after the first route table in the list of references to prevent the second route
table from short-circuiting all other routes. You can now send a request to the reviews services endpoints to get the reviews,
and all other requests to routes return the custom `404` message.
```

  namespace: gloo-mesh
spec:
  routes:
  - matchers:
    - uri:
        exact: /v1/version
    name: v1-version
    directResponseAction:
      status: 200
      body: "'v1'"
  - matchers:
    - uri:
        exact: /v2/version
    name: v2-version
    directResponseAction:
      status: 200
      body: "'v1'"
  - matchers:
    - uri:
        prefix: /
    name: not-found
    directResponseAction:
      status: 404
      body: "'custom not found'"
{{< /tab >}}
{{< tab name="CLI Inline" codelang="shell" >}}
cat << EOF | kubectl apply --context $CONTEXT_1 -f -
apiVersion: networking.enterprise.mesh.gloo.solo.io/v1beta1
kind: RouteTable
metadata:
  name: reviews-misc-rt
  namespace: gloo-mesh
  labels:
    service: reviews
spec:
  routes:
  - matchers:
    - uri:
        exact: /v1/version
    name: v1-version
    directResponseAction:
      status: 200
      body: "'v1'"
  - matchers:
    - uri:
        exact: /v2/version
    name: v2-version
    directResponseAction:
      status: 200
      body: "'v2'"
  - matchers:
    - uri:
        prefix: /
    name: not-found
    directResponseAction:
      status: 404
      body: "'custom not found'"
EOF
{{< /tab >}}
{{< /tabs >}}

> Now if you try to request the new version endpoints, you will see that you get a `404` from the reviews service as a
> reply. The reason is because the first route table that we created takes precedence due to its higher weight, which
> means the prefix route at `/reviews/v1` is getting matched before it reaches the new exact route `/reviews/v1/version`.
> You could adjust the weight on the second route table to be higher than that of the first, but that will cause the `404`
> route to then short circuit the review service routes. One solution is using a separate route table for the version
> endpoints and `404` endpoint, but there is a better way. Gloo Mesh Gateway routing supports two types of sorting. The
> first and default method is via table weight where routes are kept in the same order that they appear on their table and
> relative to table order. The alternative method is sorting by route specificity where routes are sorted via a heuristic
> that estimates how "specific" a route is and then puts more specific routes before more general ones in order to
> minimize short circuits. Exact match routes are considered more specific than regex match routes, which are considered
> more specific than prefix match routes. Routes of the same match type are than compared based on the length where
> longer matches are more specific. The sorting action is set in the `delegateAction` level and will overwrite any sorting
> done in `delegateAction`s of child `RouteTable`s. You can enable it on the `VirtualHost` to allow the routes of both
> tables to be intermingled.

Reviewer 2: ```suggestion
Now if you try to send a request to the new version endpoints, you get a `404` response from the reviews service.
The first route table that you created takes precedence due to its weight of `10`, which means that the prefix route
`/reviews/v1` is being matched before the request reaches the new exact route `/reviews/v1/version`. You could adjust the
weight of the second route table to come before the first, but the `404` route will then short-circuit the review service routes.
One solution is to use a separate route table for the version endpoints and `404` endpoint. However, an easier method
is to use the secondary type of Gloo Mesh Gateway routing.

Recall that the default type of sorting is by use of table weight, in which routes are kept in the same order that they appear
on their table and relative to table order. The alternative type of sorting is by route specificity, in which routes are
sorted according to a heuristic that estimates how specific a route is and orders routes from most to least specific in
order to minimize short circuits. Exact match routes are considered the most specific, followed by regex match routes,
followed by prefix match routes. Routes of the same match type are then compared based on length, in which longer
matches are more specific.

The sorting action is set at the `delegateAction` level, and overwrites any sorting in `delegateAction`s of child
`RouteTable`s. You can enable the `ROUTE_SPECIFICITY` sorting action on the `VirtualHost` to allow the routes of both
tables to be sorted together.
```

  namespace: gloo-mesh
spec:
  domains:
  - www.example.com
  routes:
  - matchers:
    - uri:
        prefix: /ratings
    name: ratings
    routeAction:
      destinations:
      - kubeService:
          clusterName: mgmt-cluster
          name: ratings
          namespace: bookinfo
  - matchers:
    - uri:
        prefix: /reviews
    name: reviews
    delegateAction:
      selector:
        namespaces:
        - gloo-mesh
      sortMethod: ROUTE_SPECIFICITY
{{< /tab >}}
{{< tab name="CLI Inline" codelang="shell" >}}
cat << EOF | kubectl apply --context $CONTEXT_1 -f -
apiVersion: networking.enterprise.mesh.gloo.solo.io/v1beta1
kind: VirtualHost
metadata:
  name: demo-virtualhost
  namespace: gloo-mesh
spec:
  domains:
  - www.example.com
  routes:
  - matchers:
    - uri:
        prefix: /ratings
    name: ratings
    routeAction:
      destinations:
      - kubeService:
          clusterName: mgmt-cluster
          name: ratings
          namespace: bookinfo
  - matchers:
    - uri:
        prefix: /reviews
    name: reviews
    delegateAction:
      selector:
        namespaces:
        - gloo-mesh
      sortMethod: ROUTE_SPECIFICITY
EOF
{{< /tab >}}
{{< /tabs >}}

> The resulting routes will be sorted such that the version endpoints can be hit as can the service endpoints, and all
> other requests will be caught by the `404` route.

Reviewer 2: ```suggestion
The resulting routes are sorted such that both the version endpoints and the service endpoints can be matched against, and all
other requests are matched to the `404` route.
```
